### PR TITLE
Occ config logging

### PR DESCRIPTION
**Pull request by satyakamala03**

### CODE REVIEW

File: lib/config.go [hunk 1]
```diff
@@ -461,9 +470,175 @@ func InitConfig() error {
...
+			"bind_eviction_decr_per_sec":        gAppConfig.BindEvictionDecrPerSec,
+			"bind_eviction_target_conn_pct":     gAppConfig.BindEvictionTargetConnPct,
+			"bind_eviction_max_throttle":        gAppConfig.BindEvictionMaxThrottle,
+			"bind_eviction_names":               gAppConfig.BindEvictionNames,
+			"skip_eviction_host_prefix":         gAppConfig.SkipEvictRegex,
+			"eviction_host_prefix":              gAppConfig.EvictRegex,
+			"query_bind_blocker_min_sql_prefix": gAppConfig.QueryBindBlockerMinSqlPrefix,
+			"enable_connlimit_check":            gAppConfig.EnableConnLimitCheck,
+		},
+		"SOFT-EVICTION": {
```
Comment (rasamala83) [inline lib/config.go:541]: Add a section for SATURATION_RECOVERY

saturation_recover_threshold = 30
saturation_recover_throttle_rate = 30

Log bind or soft eviction only if saturation_recover_throttle_rate > 0

File: lib/config.go [hunk 2]
```diff
@@ -461,9 +470,175 @@ func InitConfig() error {
...
+			"max_db_connects_per_sec":              gAppConfig.MaxDbConnectsPerSec,
+			"max_lifespan_per_child":               gOpsConfig.maxLifespanPerChild,
+			"max_requests_per_child":               gOpsConfig.maxRequestsPerChild,
+			"max_desire_healthy_worker_pct":        gAppConfig.MaxDesiredHealthyWorkerPct,
+		},
+		"R-W-SPLIT": {
+			"readonly_children_pct":          gAppConfig.ReadonlyPct,
+			"cfg_from_tns_override_rw_split": gAppConfig.CfgFromTnsOverrideRWSplit,
+		},
+		"STATE-LOG": {
```
Comment (rasamala83) [inline lib/config.go:563]: STATE-LOG may not require

File: lib/config.go [hunk 3]
```diff
@@ -461,9 +470,175 @@ func InitConfig() error {
...
+	//TODO: Remove log statements
+	//if logger.GetLogger().V(logger.Warning) {
+	//	logger.GetLogger().Log(logger.Warning, "Inside LogOccConfigs()")
+	//}
+	whiteListConfigs := map[string]map[string]interface{}{
+		"BACKLOG": {
+			"backlog_pct":                  gAppConfig.BacklogPct,
+			"request_backlog_timeout":      gAppConfig.BacklogTimeoutMsec,
+			"short_backlog_timeout":        gAppConfig.ShortBacklogTimeoutMsec,
+			"saturation_recover_threshold": gOpsConfig.satRecoverThresholdMs,
```
Comment (rasamala83) [inline lib/config.go:498]: Remove "saturation_recover_threshold" property from generic backlog eviction or timeout

File: lib/config.go [hunk 4]
```diff
@@ -461,9 +470,175 @@ func InitConfig() error {
...
+			s := fmt.Sprintf("%v", val)
+			//fmt.Println("interface{} to string: ", s)
+			evt.AddDataStr(cfg, s)
+		}
+		evt.Completed()
+
+		configsMarshal, _ := json.Marshal(configs)
+		configsMarshalStr := string(configsMarshal)
+
+		//TODO: remove below logs before final push
```
Comment (rasamala83) [inline lib/config.go:634]: we can validate using file-based CAL log. so we can remove this JSON marshal code and server logging

**Review by rasamala83 — CHANGES_REQUESTED** [review]
Add tests to validate this flow and periodic logging

File: lib/config.go (hunk 2 above)
Comment (satyakamala03) [reply lib/config.go:563]: removed

File: lib/config.go [hunk 5]
```diff
@@ -461,9 +473,171 @@ func InitConfig() error {
...
+
 	return nil
 }
 
+func LogOccConfigs() {
+	//TODO: Remove log statements
+	//if logger.GetLogger().V(logger.Warning) {
+	//	logger.GetLogger().Log(logger.Warning, "Inside LogOccConfigs()")
+	//}
+	whiteListConfigs := map[string]map[string]interface{}{
```
Comment (rasamala83) [inline lib/config.go:496]: change map[string]map[string]interface{} to map[string]map[string]*interface{}

File: mux/mux.go [hunk 6]
```diff
@@ -17,10 +17,9 @@
...
 package main
 
-import (
-	lib "github.com/paypal/hera/lib"
-)
+import "github.com/paypal/hera/lib"
 
 func main() {
-	lib.Run()
+	//lib.Run()
```
Comment (rasamala83) [inline mux/mux.go:23]: changes not requires in this file.

File: tests/unittest/config_logging/main_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+func TestMain(m *testing.M) {
+	//os.Exit(testutil.UtilMain(m, cfg, before))
+	racmaint = false
+	testutil.UtilMain(m, cfg, before)
+}
+
+func TestConfigLogging(t *testing.T) {
+	logger.GetLogger().Log(logger.Debug, "Test config-logging begin +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
+
+	if 4 == testutil.RegexCountFile("OCC_CONFIG", "cal.log") {
```
Comment (rasamala83) [inline tests/unittest/config_logging/main_test.go:49]: This condition always false..that is the reason test showing passed but we can't see OCC_CONFIG in CAL.
Change condition and fix the test.

File: lib/config.go (hunk 1 above)
Comment (satyakamala03) [reply lib/config.go:538]: added a section

File: lib/main.go [hunk 8]
```diff
@@ -71,29 +90,12 @@ func Run() {
...
-			evt.Completed()
-		}
-	}
-	caltxn.Completed()
-
 	//
 	// create singleton broker and start worker/pools
 	//
 	nameForTns := *namePtr
+	//InitPyplAppConfig()
```
Comment (venkatsridhar95) [inline lib/main.go:97]: Clean up these lines

File: lib/main.go [hunk 9]
```diff
@@ -50,6 +50,25 @@ func Run() {
 
 	rand.Seed(time.Now().Unix())
 
+	//
+	// worker also initialize a calclent with the same poolname using threadid==0 in
```
Comment (venkatsridhar95) [inline lib/main.go:54]: Why did we make this change? Pls revert.

File: lib/main.go [hunk 10]
```diff
@@ -153,6 +155,7 @@ func Run() {
 		lsn = NewTLSListener(fmt.Sprintf("0.0.0.0:%d", GetConfig().Port))
 	} else {
 		lsn = NewTCPListener(fmt.Sprintf("0.0.0.0:%d", GetConfig().Port))
+		//lsn = NewCryptoTLSListener(fmt.Sprintf("0.0.0.0:%d", GetConfig().Port), sec, cert)
```
Comment (venkatsridhar95) [inline lib/main.go:158]: Clean up these lines

File: lib/main.go [hunk 11]
```diff
@@ -166,6 +169,12 @@ func Run() {
 	}
 	InitRacMaint(*namePtr)
 
+	if err != nil {
```
Comment (venkatsridhar95) [inline lib/main.go:172]: err references to GetWorkerPool() or InitShardingCfg().

File: lib/config.go [hunk 12]
```diff
@@ -461,9 +466,150 @@ func InitConfig() error {
 		gAppConfig.MaxDesiredHealthyWorkerPct = 90
 	}
 
+	go func() {
+		configLoggingIntervalInHr := time.Duration(GetConfig().ConfigLoggingReloadTimeHours) * time.Hour
```
Comment (venkatsridhar95) [inline lib/config.go:470]: Let us move this to main.go and not have this here.

File: mux/mux.go [hunk 13]
```diff
@@ -17,9 +17,7 @@
 
 package main
 
-import (
-	lib "github.com/paypal/hera/lib"
-)
+import "github.com/paypal/hera/lib"
```
Comment (venkatsridhar95) [inline mux/mux.go:20]: Revert. Avoid changes that are not related to the PR.

File: lib/config.go [hunk 14]
```diff
@@ -461,9 +466,150 @@ func InitConfig() error {
...
 }
 
+func LogOccConfigs() {
+	whiteListConfigs := map[string]map[string]interface{}{
+		"BACKLOG": {
+			"backlog_pct":             gAppConfig.BacklogPct,
+			"request_backlog_timeout": gAppConfig.BacklogTimeoutMsec,
+			"short_backlog_timeout":   gAppConfig.ShortBacklogTimeoutMsec,
+		},
+		"SHARDING": {
```
Comment (venkatsridhar95) [inline lib/config.go:490]: Add bouncer as one of the features.

File: lib/config.go [hunk 15]
```diff
@@ -461,9 +466,150 @@ func InitConfig() error {
...
+			if GetSatRecoverThrottleRate() <= 0 {
+				continue
+			}
+		case "MANUAL-RATE-LIMITER":
+			if gAppConfig.EnableQueryBindBlocker == false {
+				continue
+			}
+		}
+
+		evt := cal.NewCalEvent("OCC_CONFIG", fmt.Sprintf(feature), cal.TransOK, "")
```
Comment (venkatsridhar95) [inline lib/config.go:602]: Add a sample log to the PR comment after enabling config logging.

File: tests/unittest/config_logging/main_test.go [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+	if testutil.RegexCountFile("OCC_CONFIG\tTAF", "cal.log") > 0 {
+		t.Fatalf("TAF is not enabled so we should not see TAF config logging")
+	}
+
+	if testutil.RegexCountFile("OCC_CONFIG\tR-W-SPLIT", "cal.log") > 0 {
+		t.Fatalf("All or some R-W-SPLIT config-logs are missing")
+	}
+
+	if testutil.RegexCountFile("OCC_CONFIG\tSOFT-EVICTION", "cal.log") < 1 {
+		t.Fatalf("All or some SOFT-EVICTION config-logs are missing")
```
Comment (venkatsridhar95) [inline tests/unittest/config_logging/main_test.go:144]: Why is it All or some (partial)? We are logging 1 CAL event per feature. Pls change it to 'Can't find' as that clearly states the config was not logged.

**Review by venkatsridhar95 — CHANGES_REQUESTED** [review]

**Comment by rasamala83** [comment]
E16:21:27.38	OCC_CONFIG	SATURATION-RECOVERY	0	saturation_recover_throttle_rate=30&saturation_recover_threshold=200
E16:21:27.38	LOGGER	ALERT	0	Data=alert: [PROXY workerpool.go:122] 2 randomized start ms 0
E16:21:27.38	OCC_CONFIG	WORKER-CONFIGURATIONS	0	max_stranded_time_interval=2000&high_load_max_stranded_time_interval=600111&enable_danglingworker_recovery=false&max_desire_healthy_worker_pct=90&lifespan_check_interval=10&lifo_scheduler_enabled=true&high_load_skip_initiate_recover_pct=80&max_db_connects_per_sec=10000&max_lifespan_per_child=0&max_requests_per_child=0
E16:21:27.38	OCC_CONFIG	NO-CATEGORY	0	database_type=0&cfg_from_tns=true&log_level=5&high_load_pct=130&init_limit_pct=125&num_standby_dbs=1
E16:21:27.38	LOGGER	ALERT	0	Data=alert: [PROXY workerpool.go:122] 0 randomized start ms 0
E16:21:27.38	OCC_CONFIG	SHARDING	0	enable_whitelist_test=false&sharding_cfg_reload_interval=3600&num_shards=3&scuttle_col_name=scuttle_id&max_scuttle=1024&cfg_from_tns_override_num_shards=-1&shard_key_value_type_is_string=false&whitelist_children=0&use_shardmap=true&shard_key_name=id&hostname_prefix=map[]&sharding_cross_keys_err=false&sharding_algo=false&enable_sharding=true&sharding_postfix=
E16:21:27.38	OCC_CONFIG	BIND-EVICTION	0	bind_eviction_threshold_pct=60&query_bind_blocker_min_sql_prefix=20&enable_connlimit_check=false&child.executable=mysqlworker&bind_eviction_decr_per_sec=10&bind_eviction_target_conn_pct=50&bind_eviction_max_throttle=20&bind_eviction_names=id,num,srcPrefixApp&skip_eviction_host_prefix=&eviction_host_prefix=
E16:21:27.38	OCC_CONFIG	SOFT-EVICTION	0	soft_eviction_effective_time=10000&soft_eviction_probability=50
E16:21:27.38	OCC_CONFIG	RAC	0	management_table_prefix=hera&rac_sql_interval=0&rac_restart_window=240
E16:21:27.38	OCC_CONFIG	BACKLOG	0	backlog_pct=30&request_backlog_timeout=1000&short_backlog_timeout=30
E16:21:27.38	OCC_CONFIG	BOUNCER	0	bouncer_enabled=true&bouncer_startup_delay=10&bouncer_poll_interval_ms=100

File: lib/main.go (hunk 9 above)
Comment (rasamala83) [reply lib/main.go]: Done

**Review by venkatsridhar95 — APPROVED** [review]
